### PR TITLE
[Merged by Bors] - Optimize buildVRFMessage in Hare oracle

### LINKED ISSUE #1569: Beacon may get contextually valid blocks multiple times, no check for zero blocks
1. Multiple get requests for contextually valid blocks may occur event though we have cache since we are not using a lock.

2. Should check if got no blocks (no error, but also no contextually valid blocks) and return an error - cannot verify.

### PR DESCRIPTION
**Pull request by gavraz**
## Motivation
Closes #1569

## Changes
warning when beacon gets zero contextually valid blocks
buildVRFMessage is now concurrency-optimized (added lock and cache).

## Test Plan
Added UT including a concurrency test.

### CODE REVIEW

File: hare/eligibility/oracle.go
Comment (almogdepaz) [inline hare/eligibility/oracle.go:20]: how long until @barakshani  want's this configurable ?

Comment (gavraz) [reply hare/eligibility/oracle.go:20]: the number of rounds is constant in the protocol.
the number of concurrent layers is ideally 2. It could be more if there are consensus processes running in parallel (which is not normal).
@barakshani  how big do you want the caches to be?

**Comment by gavraz** [comment]
bors merge

**Comment by gavraz** [comment]
@y0sher waiting for your review

File: hare/eligibility/beacon.go
Comment (y0sher) [inline hare/eligibility/beacon.go:69]: so maybe log this only inf not genesis ? if it's too hard to know if it's genesis then not important

File: hare/eligibility/oracle.go
Comment (y0sher) [inline hare/eligibility/oracle.go:137]: the cache itself isn't locked ?

File: hare/eligibility/oracle.go
Comment (y0sher) [inline hare/eligibility/oracle.go:264]: haven't you jut checked this above outside the lock ?

**Review by y0sher — APPROVED** [review]
I approved though I leave it to you to decide if my comments worth something :)

File: hare/eligibility/beacon.go
Comment (barakshani) [reply hare/eligibility/beacon.go:69]: If everything is fine, then in layer 2 (tortoise eats layer 1 blocks so layer 0 becomes valid) we should stop getting this message, because blocks of `sl` (should be layer 0) are valid. So only in layer 1 we expect to ignore this message.

File: hare/eligibility/oracle.go
Comment (barakshani) [inline hare/eligibility/oracle.go]: What's going on here?

File: hare/eligibility/oracle.go
Comment (barakshani) [inline hare/eligibility/oracle.go]: please write "Hare beacon"

File: hare/eligibility/beacon.go
Comment (gavraz) [reply hare/eligibility/beacon.go]: Already discussed it with Barak, said it is enough (saves code complexity).

File: hare/eligibility/oracle.go
Comment (gavraz) [reply hare/eligibility/oracle.go:137]: The cache is locked but we want to avoid double calculations.

File: hare/eligibility/oracle.go
Comment (gavraz) [reply hare/eligibility/oracle.go:264]: Yes, it is an optimization, though after rethinking it, it is probably negligible.

Comment (gavraz) [reply hare/eligibility/oracle.go:264]: I removed it from both the activesCache and the vrfMsgCache

**Comment by gavraz** [comment]
bors merge

File: hare/eligibility/beacon.go
Comment (barakshani) [reply hare/eligibility/beacon.go:69]: Looking at the function m.ContextuallyValidBlock, it looks like this case is already handled and that you can remove the parenthesis

**Pull request closed without merging**
